In [1]:
import numpy as np
import sympy as sp

# Exercise 3.1

$\xi(\omega) = \frac{1}{2} \sigma^2 - r_{xd} \omega + \frac{1}{2}r_x \omega^2$

The gradient vector $g(n) = - r_{xd} + r_x \omega$

The cost function $\xi(\omega) (n+1)$ by the steepest descent algorithm is given by:

 $\xi(\omega (n+1)) = \xi(\omega(n)) - \eta g(n)^2$. After substituting the value of $g(n)$, this is equal to:
 
 $\xi(\omega(n+1)) = \xi(\omega(n)) - \eta (- r_{xd} + r_x \omega(n))^2$
 
 = $ (\frac{1}{2} \sigma^2 - r_{xd} \omega(n) + \frac{1}{2}r_x \omega(n)^2)-  \eta (- r_{xd} + r_x \omega(n))^2$
 
 =$ (\eta r^{2}_x + \frac{1}{2}r_x)\omega(n)^2 -(2\eta r_{xd}r_x  +r_{xd}) \omega(n)+ (\eta r^{2}_{xd} + \frac{1}{2} \sigma^2) $

# Exercise 3.2

In [25]:
wo,w1 = sp.symbols("w0,w1")

w = sp.Matrix([wo,w1])
r = sp.Matrix([0.8182,0.354])
R = sp.Matrix([[1,0.8182],
               [0.8182,1]])
print "The weight vector w ="
sp.pprint(w)
print "\n r_xd = "
sp.pprint(r)
print "\nAnd R_x = "
sp.pprint(R)


The weight vector w =
⎡w₀⎤
⎢  ⎥
⎣w₁⎦

 r_xd = 
⎡0.8182⎤
⎢      ⎥
⎣0.354 ⎦

And R_x = 
⎡  1     0.8182⎤
⎢              ⎥
⎣0.8182    1   ⎦
[0.8182⋅w₀ + 0.354⋅w₁]
[w₀⋅(0.5⋅w₀ + 0.4091⋅w₁) + w₁⋅(0.4091⋅w₀ + 0.5⋅w₁)]


$\xi(\omega) = \frac{1}{2}\sigma^{2} - r_{xd}^T \omega + \frac{1}{2}\omega^T R_x \omega$.

Substituting the values of $r_{xd}$ and $R_x$, the values of $ r_{xd}^T \omega$ and $\frac{1}{2}\omega^T R_x \omega$ is :

In [32]:
rw = r.T * w
sp.pprint(rw)

[0.8182⋅w₀ + 0.354⋅w₁]


In [33]:
w_tRw = 0.5*w.T*R*w
sp.pprint(w_tRw)

[w₀⋅(0.5⋅w₀ + 0.4091⋅w₁) + w₁⋅(0.4091⋅w₀ + 0.5⋅w₁)]


Thus $\xi(\omega) = \frac{1}{2}\sigma^{2} - [0.8182⋅\omega_0 + 0.354⋅\omega_1] + [\omega_0⋅(0.5⋅\omega_0 + 0.4091⋅\omega_1) + \omega_1⋅(0.4091⋅\omega_0 + 0.5⋅\omega_1)]$

 $\xi(\omega) = \frac{1}{2}\sigma^{2} + 0.5.\omega_0^{2} + 0.5.\omega_1^{2} + 0.8182.\omega_0.\omega_1 - 0.8182.\omega_0 - 0.354.\omega_1$



In [73]:
sigma = sp.Symbol('sigma')
xi = 0.5*sigma**2 + (0.5*wo**2) + (0.5*w1**2) + (0.8182*wo*w1) - (0.8182*wo) - (0.354*w1)
eqn1 = sp.diff(xi,wo)
eqn2 = sp.diff(xi,w1)
gradient = sp.Matrix([eqn1,eqn2])
print "Gradient :"
sp.pprint(gradient)

Gradient :
⎡1.0⋅w₀ + 0.8182⋅w₁ - 0.8182⎤
⎢                           ⎥
⎣0.8182⋅w₀ + 1.0⋅w₁ - 0.354 ⎦


In [74]:

w_optimal = sp.solve([eqn1,eqn2],[wo,w1])
print "The optimal weight vector = ",w_optimal

The optimal weight vector =  {w0: 1.59902944424901, w1: -0.954325891284541}


# Part (b)

In [88]:
def steepest_descent(eta, xi , gradient):
    w_init = np.array([0,0])
    
    w_n = w_init.T
    xi_n = xi.subs({wo:w_n[0],w1:w_n[1]})
    gradient_n = gradient.subs({wo:w_n[0],w1:w_n[1]})
    for i in range(200):
        eta_gn  =  np.asarray(eta*gradient_n)
        w_new = w_n - np.array([eta_gn[0,0],eta_gn[1,0]])
        #Update the weight vector, the gradient and corresponding value of xi
        
        gradient_n = gradient.subs({wo:w_new[0],w1:w_new[1]})
        xi_new = xi.subs({wo:w_new[0],w1:w_new[1]})
        
        xi_n = xi_new
        w_n = w_new
        
    return w_n    

In [90]:
sp.pprint(steepest_descent(0.3,xi,gradient))


[1.59901227650625 -0.954308723541781]


In [91]:
sp.pprint(steepest_descent(1.0,xi,gradient))

[1.59902944424901 -0.954325891284542]
